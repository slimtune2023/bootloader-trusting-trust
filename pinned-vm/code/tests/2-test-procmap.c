// simple example of using procmap.  replicates <1-test-one-addr.c>
#include "rpi.h"
#include "procmap.h"

void notmain(void) { 
    kmalloc_init_set_start((void*)SEG_HEAP, MB(1));

    // read write the first mb.
    uint32_t user_addr = MB(16);
    assert((user_addr>>12) % 16 == 0);
    uint32_t phys_addr1 = user_addr;
    PUT32(phys_addr1, 0xdeadbeef);

    procmap_t p = procmap_default_mk(dom_kern);
    procmap_push(&p, pr_ent_mk(user_addr, MB(1), MEM_RW, dom_user));

    trace("about to enable\n");
    procmap_pin_on(&p);

    lockdown_print_entries("about to turn on first time");
    assert(mmu_is_enabled());
    trace("MMU is on and working!\n");
    
    uint32_t x = GET32(user_addr);
    trace("asid 1 = got: %x\n", x);
    assert(x == 0xdeadbeef);
    PUT32(user_addr, 1);

    pin_mmu_disable();
    assert(!mmu_is_enabled());
    trace("MMU is off!\n");
    trace("phys addr1=%x\n", GET32(phys_addr1));
    assert(GET32(phys_addr1) == 1);

    trace("SUCCESS!\n");
}
