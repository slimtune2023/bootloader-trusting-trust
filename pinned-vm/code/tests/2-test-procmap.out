TRACE:notmain:about to enable
TRACE:lockdown_print_entries:-----  <about to turn on first time> ----- 
TRACE:lockdown_print_entries:  pinned TLB lockdown entries:
TRACE:lockdown_print_entry:   idx=0
TRACE:lockdown_print_entry:     va_ent=0x20000200: va=0x20000|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x200000c3: pa=0x20000|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x80: dom=1|xn=0|tex=0|C=0|B=0
TRACE:lockdown_print_entry:   idx=1
TRACE:lockdown_print_entry:     va_ent=0x20100200: va=0x20100|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x201000c3: pa=0x20100|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x80: dom=1|xn=0|tex=0|C=0|B=0
TRACE:lockdown_print_entry:   idx=2
TRACE:lockdown_print_entry:     va_ent=0x20200200: va=0x20200|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x202000c3: pa=0x20200|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x80: dom=1|xn=0|tex=0|C=0|B=0
TRACE:lockdown_print_entry:   idx=3
TRACE:lockdown_print_entry:     va_ent=0x200: va=0x0|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0xc3: pa=0x0|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=4
TRACE:lockdown_print_entry:     va_ent=0x100200: va=0x100|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x1000c3: pa=0x100|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=5
TRACE:lockdown_print_entry:     va_ent=0x8f00200: va=0x8f00|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x8f000c3: pa=0x8f00|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=6
TRACE:lockdown_print_entry:     va_ent=0x7f00200: va=0x7f00|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x7f000c3: pa=0x7f00|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x88: dom=1|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entry:   idx=7
TRACE:lockdown_print_entry:     va_ent=0x1000200: va=0x1000|G=1|ASID=0
TRACE:lockdown_print_entry:     pa_ent=0x10000c3: pa=0x1000|nsa=0|nstid=0|size=11|apx=1|v=1
TRACE:lockdown_print_entry:     attr=0x108: dom=2|xn=0|tex=1|C=0|B=0
TRACE:lockdown_print_entries:----- ---------------------------------- 
TRACE:notmain:MMU is on and working!
TRACE:notmain:asid 1 = got: 0xdeadbeef
TRACE:notmain:MMU is off!
TRACE:notmain:phys addr1=0x1
TRACE:notmain:SUCCESS!
